### ISSUE TITLE
Passing decimals as amount

### ISSUE THREAD

**Issue by sarneeh** [dinerojs/dinero.js#58]
**Is your feature request related to a problem? Please describe.**

So if you would like to integrate with services like [Braintree](https://developers.braintreepayments.com/reference/request/subscription/create/ruby#add_ons.add[#].amount) (or any other service that uses decimals instead of minor currency values) you will have to manually convert those decimals/strings to integers just to make it compatible with Dinero. 

**Describe the solution you'd like**

It would be great if I could just pass a string like `499.99` into Dinero and it would convert it into `49999`. This could be achieved for example with a `fromFormat` function which would take a string mask argument and a value: `fromFormat('0.00', '499.99')`.

**Describe alternatives you've considered**

Manual conversion.

**Comment by sarahdayan** [comment]
Hey @sarneeh and thanks for the suggestion. You're not the first one to ask and it could indeed be interesting to add this, maybe as a static method or as a separate factory (to keep the bundle size untouched for those who don't need it).

I guess it could auto-detect the precision based on the number of digits after the decimal point, which would be even simpler. Something like:

```js
DineroFactory.fromFloat({ amount: 499.99, currency: 'EUR' })
```

Would that help?

**Comment by sarneeh** [comment]
@sarahdayan Indeed! This idea is a lot better as it keeps the Dinero API consistent! That would definitely help :smiley:

**Comment by sarahdayan** [comment]
@sarneeh In this case we can share the load and get this out sooner: I can implement it and you could review (should demand less time than implementing). Would this be doable for you?

**Comment by sarneeh** [comment]
@sarahdayan I'll find time for a review for sure! Thank you!

**Comment by jadamduff** [comment]
Very excited about this feature. I'd love to be able to pass in strings and have them parsed into the object. Ex: "$150" -> {amount: 15000}

**Comment by sarahdayan** [comment]
@jadamduff This would be the job of another factory, such as `fromString`. But this involves a lot more work.

**Comment by jadamduff** [comment]
@sarahdayan okay thank you

**Comment by captDaylight** [comment]
I'd be interested in taking this on @sarahdayan if you haven't started already.

**Comment by fish4j** [comment]
Can we simply use `Dinero({ amount: Math.round(499.99 * Math.pow(10,minor unit)) })`?

**Comment by sarahdayan** [comment]
@captDaylight On it already, but thanks for offering your help! Would love it if you want to review.
@fish4j Yes, this is a factory that you can definitely build for yourself but which involves that you pass the precision (minor unit in your snippet) by hand. The idea behind `fromFloat` is to detect it.

**Comment by blacktrue** [comment]
Hi @sarahdayan 

The factory with fromFloat/fromString methods is already available? I do not find anything about it

Thanks

**Comment by sarahdayan** [comment]
@blacktrue Hey, I'm still working on it, but got delayed 🙂 If you subscribe to notifications for this issue you'll get pinged once the PR that fixes it is merged and the version that embarks it is deployed.

**Comment by ishegg** [comment]
@sarahdayan hey, are you in need of a helping hand? If you're swamped with work I might be able to give it a try.

From a first thought, it seems that something like

```
fromFloat(amount, currency) {
  var precision = amount.split('.')[1].length;
  var amount = Math.round(Number(amount) * Math.pow(10, precision));
  return Dinero({
    amount: amount,
    currency: currency,
    precision: precision
  })
}
```

in `services/static.js` would do it. (This is missing pre/postcondition assertions)

**Comment by sarahdayan** [comment]
Hey @ishegg, thanks for offering your help! I'm indeed swamped with work, had to put this on the back burner. Feel free to PR 🙂

**Comment by sarahdayan** [comment]
Hey! I just published a RFC for Dinero.js v2 🚀 It tackles this specific problem.

I would love to have your feedback on it: https://github.com/dinerojs/rfcs/pull/1

cc @ishegg @sarneeh @jadamduff @captDaylight @fish4j

**Comment by Huespal** [comment]
I am looking exactly for this. When will be released? Thanks.

**Comment by bugproof** [comment]
wouldn't it be better to integrate some other library like [bignumber.js or decimal.js or big.js? ](https://github.com/MikeMcl/big.js/wiki#what-is-the-difference-between-bigjs-bignumberjs-and-decimaljs)

**Comment by sarneeh** [comment]
@bugproof You probably don't want to use multiple money formatting libraries, as it's only adds unnecessary complexity. The case of this issue is to be able to more easily integrate `dinero` with different money formats used in available payment gateway (or other) APIs.

**Comment by sarahdayan** [comment]
@Huespal I'm currently not able to tackle this because of work, so I can't give you an ETA. Feel free to PR if you need it ASAP, I can take time to review.

@bugproof There are several things to consider. First, it's much better to use non-floats (as `number` or `bigint`) because they're more reliable and performant when it comes to performing math on them. Dinero used `number` so far because `bigint` weren't available, but now that they are it's going to become an option in v2. Secondly, the point of this issue is to create factories so you can provide a format that is convenient for you. It wouldn't change the internals of Dinero, and what type(s) it uses. Finally, regarding libraries, I definitely don't want to create any coupling with a third-party, nor use any as a dependency. I think Dinero should stay dependency-free, and as agnostic as possible. However, writing adaptors so you can provide entities created with third-parties could be an idea, and Dinero v2 could definitely provide an API for this.

**Comment by bugproof** [comment]
it's problematic when you have API that returns decimal values like 123.32 how can you pass it as amount to dinero? multiply by 100?

**Comment by johnnyperdomo** [comment]
Any updates on this?

**Comment by canercak** [comment]
Hi , I just discovered dinero.js and I think this issue is very necessary because most of the time money is used with text-masking and the value returned from masking doesn't come in cents. 
For example currently I use react text-mask addon and I now give 5000 cents Dinero object as input value and still get back $50.00 as string as I got before using Dinero. My workaround is still to remove the currency, then parse the float, then multiply by 100 (while controlling the minimum sub unit for the currency). 

I think these are necessary..I will go on doing these in my fork if I can
```
DineroFactory.fromFloat({ amount: 499.99, currency: 'EUR' }) 
DineroFactory.fromString({ amount: "499.99", currency: 'EUR'  })
```
this is 'nice to have'
`DineroFactory.fromCurrencyString({ amount: "$499.99", currency: 'USD' })`

**Comment by nickpalmer** [comment]
Why not allow a `{ unit: 499.99 }` on construction that does:
```
  if (o.unit) {
    o.amount = Math.round(o.unit * Math.pow(10, o.precision || defaultPrecision));
    delete o.unit;
  }
```
Users of `{ unit: xxx }` are going to be aware that they are coming from a representation that is imprecise, and rounding is likely to occur.

**Comment by sarahdayan** [comment]
Hey! I just released [Dinero.js v2 in alpha](https://v2.dinerojs.com/docs). It doesn't provide decimal support out of the box, but I've written a [guide](https://v2.dinerojs.com/docs/faq/how-can-i-create-dinero-objects-from-floats) on how to do it userland. The new design makes it straightforward to extend the library.

I would still recommend being cautious with floats, and make sure what you're passing is really the amount you intended.

**Comment by ScreamZ** [comment]
Looks like I'm coming at perfect timing ✌🏻
